wheel/python/clvm_rs/program.py

```diff
@@ -0,0 +1,346 @@
+        assert Program.to(17) == at(p1, "rrrfrf")
+        ```
+
+        Returns `None` if an atom is hit at some intermediate node.
+
+        ```
+        p1 = Program.to(10)
+        assert None == at(p1, "rr")
+        ```
+        """
+        r = at(self, position)
+        if r is None:
+            return r
+        return self.to(r)
+
+    def at_many(self, *positions: str) -> List[Optional["Program"]]:
+        """
+        Call `.at` multiple times.
+
+        Why? So you can write
+
+        `
+        if p.at_many("f", "rf", "rfrf") == [5, 10, 15]:
+        `
+        instead of
+        `
+        if [p.at("f"), p.at("rf"), p.at("rfrf")] == [5, 10, 15]:
+        `
+        """
+        return [self.at(_) for _ in positions]
+
+    def replace(self, **kwargs) -> "Program":
+        """
+        Create a new program replacing the given paths (using `at` syntax).
+        Example:
+        ```
+        >>> p1 = Program.to([100, 200, 300])
+        >>> print(p1.replace(f=105) == Program.to([105, 200, 300]))
+        True
+        >>> p2 = [100, 200, [301, 302]]
+        >>> print(p1.replace(rrf=[301, 302]) == Program.to(p2))
+        True
+        >>> p2 = [105, 200, [301, 302]]
+        >>> print(p1.replace(f=105, rrf=[301, 302]) == Program.to(p2))
+        True
+        ```
+
+        This is a convenience method intended for use in the wallet or
+        command-line hacks where it would be easier to morph elements
+        of an existing clvm object tree than to rebuild one from scratch.
+
+        Note that `Program` objects are immutable. This function returns a
+        new object; the original is left as-is.
+        """
+        return self.to(replace(self, **kwargs))
+
+    def tree_hash(self) -> bytes:
+        # we operate on the unwrapped version to prevent the re-wrapping that
+        # happens on each invocation of `Program.pair` whenever possible
+        if self._cached_sha256_treehash is None:
+            self._cached_sha256_treehash = sha256_treehash(self._unwrapped)
+        return self._cached_sha256_treehash
+
+    def run_with_cost(
+        self, args, max_cost: int = MAX_COST, flags: int = 0
+    ) -> Tuple[int, "Program"]:
+        prog_bytes = bytes(self)
+        args_bytes = bytes(self.to(args))
+        try:
+            cost, lazy_node = run_serialized_chia_program(
+                prog_bytes, args_bytes, max_cost, flags
+            )
+            r = self.wrap(lazy_node)
+        except ValueError as ve:
+            raise EvalError(ve.args[0], self.wrap(ve.args[1]))
+        return cost, r
+
+    def run(self, args) -> "Program":
+        cost, r = self.run_with_cost(args, MAX_COST)
+        return r
+
+    def curry(self, *args) -> "Program":
+        """
+        Replicates the curry function from clvm_tools, taking advantage of *args
+        being a list.  We iterate through args in reverse building the code to
+        create a clvm list.
```

Comment:
this comment looks like it describes an earlier implementation. is `args` really required to be a list here? (if so, it should be typed as such)

Reply:
I'm a bit flummoxed as to how typing works for `*args` in python. Let me look it up.

Reply:
Okay, I want `*args: CastableType` here (as opposed to what I originally thought, which was `List[CastableType]`).